docs/adr/0001-update-logins-api.md

```diff
@@ -0,0 +1,104 @@
+* Establishing a standard encyrption approach across components
+
+
+## Considered Options
+
+* Option 1 - Reduce the API functions that require the encryption key and pass the key to the remaining functions
+* Option 2 - Implement a different key management approach
+
+
+## Decision Outcome
+
+(I'm purposely leaving this blank until we arrive at a decision.)
+
+
+## Pros and Cons of the Options
+
+### Option 1 - Reduce the API functions that require the encryption key and pass the key to the remaining functions
+
+* Description
+
+  Currently the below logins API functions would require the per-field encryption key:
+    - [add](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L362)
+    - [update](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L611)
+    - [get_all](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L273)
+    - [get_by_base_domain](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L279)
+    - [get_by_id](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L330)
+    - [check_valid_with_no_dupes](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L663)
+    - [potential_dupes_ignoring_username](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L721)
+    - [import_multiple](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L455)
+
+      **Note:** Functions related to sync have been omitted as it is assumed they will have access to decrypted data.
+
+  Propsed changes:
+    - Combine the `add` and `update` functions into a new `add_or_update` function
+      - This will allow the removal of consumer code that distinguishes when a login record should be created or updated
+    - Pass the per-field encryption key to the `import_multiple` function
+      - This function will be removed once the Fennec to Fenix migration period ends
+    - Remove both the `potential_dupes_ignoring_username` and `check_valid_with_no_dupes` from the API
+      - Neither function is called in Firefox iOS
+      - Android Components uses both to provide validation and de-duping before logins are added or updated so we can eliminate the need to externalize these functions by replicating this logic in the new `add_or_update` function
+    - Create a `decrypt_and_fixup_login` function that both decrypts a login _and_ performs the validate and fixup logic
+      - This will eliminate the need for the `get_all`, `get_by_base_domain`, and `get_by_id` API functions to perform the fixup logic
```

Comment:
I like this idea a lot, but how would it work?  Would `decrypt_and_fixup_login()` return the plaintext username/password alongside the fixed up login?  Would it store the decrypted values inside the `Login` struct?  And how would `encrypt_and_fixup_login()` work?

What about having 2 types: `Login` and `EncryptedLogin`:
- The `get_*` methods return `EncryptedLogin` instances
- We define `EncryptedLogin.decrypt_and_fixup(key) -> Login` and `Login.encrypt_and_fixup(key) -> EncryptedLogin`
- `add_or_update()` inputs an `EncryptedLogin`

But then I also wonder if we need expose `EncryptedLogin` at all.  Are consumers going to want to use logins without accessing the username/password and without going through validation?  If not, then the API can be simplified by always giving the consumer decrypted logins and requiring the key in the `get_*` methods and `save_or_update()`.   It doesn't simplify the key management for the consumer if can get logins without a key but then they always decrypt them afterwards.

Another option would be to make the main API calls return `Logins`, but then have some specialized functions that return `EncryptedLogin`.  This would be good if the consumers sometimes, but rarely, needed to use logins without a key.

Reply:
> We define EncryptedLogin.decrypt_and_fixup(key) -> Login and Login.encrypt_and_fixup(key) -> EncryptedLogin

As a purely technical note, uniffi won't currently let you define methods like that on a `dictionary` type, it would have to be either a free-standing function or a method on the `interface` type for `Store`.

IIRC, the reason we want to let consumers work with `EncryptedLogin` records is because prompting for the encryption key is a disruptive user event. So for example, if the user is on a page, we want to be able to check if we have any saved logins for that page before we prompt the user to unlock their keystore and actually decrypt the record.

Reply:
Mark explained that to me, and I'm finally 100% convinced.  From an ADR-perspective, I'll just note that you could have methods for those checks that didn't return logins.  Something like `get_potential_logins_count()`.  The reason for doing that is to avoid any potential confusion on the consumer code side with mixing up the 2 types of logins.  Maybe there would be exceptions when trying to get a plaintext data from an encrypted login or maybe the consumers would decide to hold on to the keys longer than they needed to just in case they ever needed to decrypt a login.

However, I'm realizing that the snippet you pasted above feels very natural and avoids all the issues.  You can also extend it to handling multiple logins:

```
enc_logins = store.get_by_base_domain("foo.com")
if user_wants_to_use_any_login(enc_logins):
    key = get_the_key_from_somewhere()
    logins = [store.decrypt_and_fixup(key, enc_login) for enc_login in enc_logins]
    present_login_selector(logins)
```